Test improvements; add server.CloseConnection method

daveyarwood opened the pull request:
A small part of #37. On @glynternet 's suggestion, I'm turning that big PR into a series of smaller PRs that are easier to review and merge.

daveyarwood commented:
See https://github.com/hypebeast/go-osc/pull/39/commits/52238bdd594cf3e2a6ef1bd3eaf2b26a16b3ae06 for a summary of the test improvements.

glynternet commented on osc/osc_test.go:
```diff
@@ -111,53 +113,80 @@ func TestAddMsgHandlerWithInvalidAddress(t *testing.T) {
-			client.Send(msg)
+			if err := client.Send(msg); err != nil {
+				t.Error(err)
+				done.Done()
```
These aren't required as we're returning straight away anyway.
The test is unchanged without them.

daveyarwood replied:
I was wondering about that. Do `t.Error` and `t.Errorf` halt execution?

glynternet replied:
Actually, scrap what I said, I forgot it was in a go routine, so without using `t.Fatal` here the test won't end if there is an error.

* `t.Error` is to continue on with the rest of the test logic.
* `t.Fatal` will end a test immediately.

I'm not 100% sure which would be preferred.
I think I'd prefer to `t.Fatal` and end the test immediately if there is an error sending the message, but I can see arguments for trying to recover and finish the flow.

daveyarwood replied:
I think I'm inclined to leave it alone for now, but I'm happy to change it to `t.Fatal` if @hypebeast wants.

hypebeast replied:
I'm fine with it.

glynternet commented on osc/osc_test.go:
```diff
@@ -111,53 +113,80 @@ func TestAddMsgHandlerWithInvalidAddress(t *testing.T) {
+	addr := "localhost:" + strconv.Itoa(port)
+
+	server := &Server{Addr: addr, Dispatcher: NewStandardDispatcher()}
+	defer server.CloseConnection()
```
There's no need for this call with the way you are closing the connection inside the handler.
We would end up doing it twice and not checking the response value.

daveyarwood replied:
Good point! I'll remove this line and handle the error in the place where I'm calling `server.CloseConnection()` further down.

glynternet commented on osc/osc_test.go:
```diff
@@ -111,53 +113,80 @@ func TestAddMsgHandlerWithInvalidAddress(t *testing.T) {
+				finish <- true
+			}()
 
-		d := NewStandardDispatcher()
```
Any reason for moving away from assigning `NewStandardDispatcher()` to `d` and then adding a handler?
I'm wondering if it may be more obvious to less-proficient members of the Go community whether it would be best to keep it as it previously way and avoid the type assertion.

I feel like with tests usually it's best to optimise for clarity.

daveyarwood replied:
The trouble I ran into is that the message handler function needs to have a reference to `server`, so `server` needs to be defined first.

I just realized that I can solve both problems, though, by doing it this way:

```diff
diff --git a/osc/osc_test.go b/osc/osc_test.go
index a979eb1..1b19ed0 100644
--- a/osc/osc_test.go
+++ b/osc/osc_test.go
@@ -139,9 +139,10 @@ func TestServerMessageDispatching(t *testing.T) {
 	port := 6677
 	addr := "localhost:" + strconv.Itoa(port)
 
-	server := &Server{Addr: addr, Dispatcher: NewStandardDispatcher()}
+	d := NewStandardDispatcher()
+	server := &Server{Addr: addr, Dispatcher: d}
 
-	if err := server.Dispatcher.(*StandardDispatcher).AddMsgHandler(
+	if err := d.AddMsgHandler(
 		"/address/test",
 		func(msg *Message) {
 			defer func() {
```

I wasn't sure if that would work, but it totally does, because `server.Dispatcher` is just a pointer to `d`.

glynternet replied:
Can we close the server after we send the message from the client anyway?

daveyarwood replied:
No, because that would be subject to a race condition. It's possible that we would send the message from the client, and then close the server before the server has a chance to receive the message.

glynternet replied:
I feel like that's what the WaitGroup should be protecting. We would ensure all of the logic for the test is completed and then close the server once we're happy or unhappy with the results. 

We could assign the server to a variable and then on the next line create a `defer` function that closes the server. Then whichever route out of the test function, we ensure we are always closing the server last.

Either way, this isn't something that would block this PR, from my perspective.

daveyarwood replied:
I took a closer look at this just now and you're totally right! I've fixed it up so that we just do a `defer` function right after we define the server, that closes the server.

glynternet commented on osc/osc.go:
```diff
@@ -568,6 +572,18 @@ func (s *Server) Serve(c net.PacketConn) error {
 
+// CloseConnection forcibly closes a server's connection.
+//
+// This causes a "use of closed network connection" error the next time the
```
I have a feeling I found a way around this before but I can't remember it now 🤔 
I think it involved using the read timeout value, but I wouldn't worry about working it out for the save of the tests.

glynternet commented:
Pretty much LGTM.
Just a couple of simplifications.

glynternet commented:
Also, btw, disclaimer, I have absolutely not control when it comes to this repository.
I'm just trying to lift a little support/maintenance burden.

daveyarwood commented:
Thanks for your feedback and for helping to move things along more smoothly! I've changed a few things based on your feedback and fixed the CI failure, which appears to be from an interesting recent addition to `go vet` on the master branch version of Go.
